Finishing up the documentation update

VictorBarbosaMartins opened the pull request:
Finishing up the documentation update started by #399.
- Update the docstrings from the modules subdirectories

GernotMaier reviewed (approved):
All good, thank you!

VictorBarbosaMartins commented:
> All good, thank you!

Thanks but there are quite some files to update still. I hope to be done by tomorrow, I will then request the review again.

GernotMaier commented on simtools/configuration/commandline_parser.py:
```diff
@@ -12,12 +12,18 @@ class CommandLineParser(argparse.ArgumentParser):
     Command line arguments should be given in snake_case, e.g. `input_meta`.
 
-    Methods
+    Atributes
```
Remember that numpy requires the underlines to be exactly the same length as the title
```
Attributes
----------
```

GernotMaier replied:
Should be `Attributes`

GernotMaier commented on simtools/configuration/commandline_parser.py:
```diff
@@ -12,12 +12,18 @@ class CommandLineParser(argparse.ArgumentParser):
+    db_config: bool
+        Add database configuration parameters to list of args.
+    job_submission: bool
+        Add job submission configuration parameters to list of args.
```
All these are not parameters of the class init. Could you explain why you want to move this here?

GernotMaier commented on simtools/configuration/commandline_parser.py:
```diff
@@ -31,20 +37,6 @@ def initialize_default_arguments(
     ):
         """
         Initialize default arguments used by all applications (e.g., verbosity or test flag).
```
I disagree with removing the parameter description in the method docstring.

GernotMaier commented on simtools/configuration/commandline_parser.py:
```diff
@@ -215,9 +183,13 @@ def initialize_job_submission_arguments(self):
-        Initialize default arguments for site and telescope model
-        definition
-
+        Initialize default arguments for site and telescope model definition
```
Add a blank line between summary and Parameters.

GernotMaier commented:
I think I misunderstood the PR (and didn't notice that it is still a draft). So what exactly is the plan?

Removing the list of methods is ok (it is also discussed in the numpy doc documentation as optional, see [here](https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring).

However I do think that the docstring of every class method should include the list of parameters relevant for this method. If this leads to repetition, than we should rather discuss if we need to change the code (the reason for the repetition) and not remove the documentation.

GernotMaier reviewed (changes requested):
See also comment in the main discussion.

GernotMaier commented:
OK - but in the files submitted you've removed the parameter description in a docstring and moved parameter descriptions from modules to the class description (see my comments above).

VictorBarbosaMartins commented:
So, I marked the PR ready for review. It is quite a lot of changes, but in general I:
- removed methods and attribute lists from the classes and kept only the parameter lists;
- standardised the module docstrings;
- added docstring where we were missing (including classes for exceptions);
- Added raises and returns statments where we were missing
- added all classes in a file to the `__all__` list;
- added an `__init__.py` in each repository which were missing in order to characterise it as a package (although not mandatory for the current versions of python, it is still highly recommended).

Probably I missed somethings but for now I think this will increase quite a lot the documentation coverage and let's keep the documentation standardised.

GernotMaier commented:
Making the docs gives this error message:
```
WARNING: autodoc: failed to import module 'visualize'; the following exception was raised:
No module named 'visualize'
```
Can you check?

(Remember to do often enough `make clean` before `make html` to see all the error messages.

orelgueta commented:
> Making the docs gives this error message:
> 
> ```
> WARNING: autodoc: failed to import module 'visualize'; the following exception was raised:
> No module named 'visualize'
> ```
> 
> Can you check?
> 
> (Remember to do often enough `make clean` before `make html` to see all the error messages.

This error is from my PR of moving visualize. What's strange is that I didn't see this warning locally (I remembered to clean this time), but I see that the CI showed this warning. I will try to fix it in this branch as part of the review.

GernotMaier commented:
> > Making the docs gives this error message:
> > ```
> > WARNING: autodoc: failed to import module 'visualize'; the following exception was raised:
> > No module named 'visualize'
> > ```
> > 
> > 
> >     
> >       

> > Can you check?
> > (Remember to do often enough `make clean` before `make html` to see all the error messages.
> 
> This error is from my PR of moving visualize. What's strange is that I didn't see this warning locally (I remembered to clean this time), but I see that the CI showed this warning. I will try to fix it in this branch as part of the review.

It shows up for me locally, not sure what is going on. But fixed now, thanks!

GernotMaier commented on simtools/simtel/simtel_runner.py:
```diff
@@ -173,13 +181,25 @@ def run(self, test=False, force=False, input_file=None, run_number=None):
 
     @staticmethod
     def _simtel_failed(sys_output):
+        """Test if output is different than 0
```
I think it tests if the simtel process ended successfully.

VictorBarbosaMartins replied:
I corrected it now

GernotMaier commented on simtools/simtel/simtel_runner.py:
```diff
@@ -173,13 +181,25 @@ def run(self, test=False, force=False, input_file=None, run_number=None):
-        Raise sim_telarray execution error. Final 30 lines from the log file
-        are collected and printed.
+        Raise sim_telarray execution error. Final 30 lines from the log file are collected and
+        printed.
```
Do we need a backslash here to connect lines?

VictorBarbosaMartins replied:
For comments it does seem to be necessary but for consistency I am adding one here

GernotMaier commented on simtools/simtel/simtel_runner.py:
```diff
@@ -197,6 +217,11 @@ def _raise_simtel_error(self):
+        Raises
+        ------
+        SimtelExecutionError
+            if output of command is 0.
```
Maybe better to speak about success or failure here, as we don't test for output `0` here.

GernotMaier commented on simtools/util/names.py:
```diff
@@ -42,19 +54,21 @@ def validate_camera_name(name):
     str
         Validated name.
+
+    Raises
```
Technically, this is not correct. This method does not raise any exceptions (it is in `validate_name`). Comments applies to all similar methods.

VictorBarbosaMartins replied:
Ok, I thought we could propagate it. I guess this would be the case, when the Raise statement is inside a private function, which does not make it to the API, right? Anyway, it is not the case here. I removed the misplaced Raises.

GernotMaier replied:
I don't see the changes, but let's leave it. Propagating means would mean what? One level, two levels, up to the applications - I guess you see the issue. Best to document where it happens, and that's it.

GernotMaier reviewed (changes requested):
Huge effort with significant improvements.

There are a couple of comments to be address and I will try to summarize more general things in the comment sections.

I've fixed typos I've found directly.

GernotMaier commented:
Here some general remarks or items to be discussed:

- we are not not following the syntax of sphinx anymore (correct me if I am wrong):
```
   off_axis: (float, required)
        Off-axis angle in deg.
```
Ssee https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes , I think we should follow this approach. Now the question is: Victor put such a huge effort into this and changed it everywhere, Is there an easy way (maybe a `sed` command) to resolve this?

- minimal or full description of parameters?
```\
label: (str, optional)
            Instance label. Important for output file naming (default is None).
```
or
```\
label: str
            Instance label. Important for output file naming.
```
The full description includes duplication which requires to sync any changes of defaults with the docstring (especially the default values).

- single paragraph or single sentences or combined into one line?
```
    def _read_workflow_configuration(self, workflow_config_file):
        """
        Read workflow configuration from file and merge it with existing workflow config. Keys are \
         changed to lower case.
```
or
```
    def _read_workflow_configuration(self, workflow_config_file):
        """
        Read workflow configuration from file and merge it with existing workflow config. 
        
        Keys are changed to lower case.
```
My personal reference is to separate those sentences for improved readability.

More minor:

- are we removing class Attributes sections or now? It is a bit confusing right now.
-  capital letters at the beginning of a sentence? It is inconsistent between the docstrings.
- a `.` at the end of a sentence or not?
- for the future: I think we should slowly start introduce class or method docstrings. Sometimes hard to see what a module is doing (e.g., mode/model_utils)

orelgueta commented on simtools/configuration/commandline_parser.py:
```diff
@@ -33,18 +34,17 @@ def initialize_default_arguments(
+        paths: (bool, optional)
+            Add path configuration to list of args (default is True).
+        telescope_model: (bool, optional)
+            Add telescope model configuration to list of args (default is False).
```
I vote for not mentioning in the documentation which parameters are optional and which aren't. I agree that it is essentially duplication and that very quickly we will forget to write it and then there will be a mismatch.

VictorBarbosaMartins replied:
I now removed the duplication and adjusted it to sphinx syntax

orelgueta commented on simtools/data_model/workflow_description.py:
```diff
@@ -211,20 +197,18 @@ def product_data_file_format(self, suffix=False):
-            if relevant fields are not defined in top level metadata
-            dictionary
-
+            if relevant fields are not defined in top level metadata dictionary.
```
I would vote to always write actual text. I.e., start sentences with capital letters, `.` at the end of sentences and full explicit descriptions.

VictorBarbosaMartins replied:
I am with Orel, though going through the whole documentation correcting for `.` at the end of sentences would be rather impractical

orelgueta replied:
Definitely impractical, but we can try doing it in the same way we plan to increase coverage. I.e., if we see a sentence that needs correcting while working on other things, then we correct it.

VictorBarbosaMartins commented on simtools/configuration/commandline_parser.py:
```diff
@@ -58,16 +58,14 @@ def initialize_default_arguments(
-            gammasim-tools configuration file.
         workflow_config (str, optional)
-            workflow configuration file.
+            workflow configuration file (default is False).
```
Ok, now I removed those statements as well.

VictorBarbosaMartins commented on simtools/corsika/corsika_config.py:
```diff
@@ -47,39 +51,35 @@ class CorsikaConfig:
+    primary: int
+        CORSIKA particle ID
+    site: str
+        North or South.
```
I agree. I just removed the attributes sections.

VictorBarbosaMartins commented:
> Here some general remarks or items to be discussed:
> 
> * we are not not following the syntax of sphinx anymore (correct me if I am wrong):
> 
> ```
>    off_axis: (float, required)
>         Off-axis angle in deg.
> ```
> 
> Ssee https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes , I think we should follow this approach. Now the question is: Victor put such a huge effort into this and changed it everywhere, Is there an easy way (maybe a `sed` command) to resolve this?
> 
> * minimal or full description of parameters?
> 
> ```
> label: (str, optional)
>             Instance label. Important for output file naming (default is None).
> ```
> 
> or
> 
> ```
> label: str
>             Instance label. Important for output file naming.
> ```
>

I corrected and adjusted all the code in simtools to sphinx. Please let me know if something is missing.

> The full description includes duplication which requires to sync any changes of defaults with the docstring (especially the default values).
> 
> * single paragraph or single sentences or combined into one line?
> 
> ```
>     def _read_workflow_configuration(self, workflow_config_file):
>         """
>         Read workflow configuration from file and merge it with existing workflow config. Keys are \
>          changed to lower case.
> ```
> 
> or
> 
> ```
>     def _read_workflow_configuration(self, workflow_config_file):
>         """
>         Read workflow configuration from file and merge it with existing workflow config. 
>         
>         Keys are changed to lower case.
> ```
> 
> My personal reference is to separate those sentences for improved readability.

Ok, let's keep sentences separated for better readability.

> 
> More minor:
> 
> * are we removing class Attributes sections or now? It is a bit confusing right now.

Yes, I removed the ones that I forgot before.

> * capital letters at the beginning of a sentence? It is inconsistent between the docstrings.
> * a `.` at the end of a sentence or not?

I prefer to have `.` at the end of sentences, yes. But I prefer to correct it as we go and not search for it throughout the documentation now.

> * for the future: I think we should slowly start introduce class or method docstrings. Sometimes hard to see what a module is doing (e.g., mode/model_utils)

I am not sure I understand this point. We already write the docstrings of methods and classes.

GernotMaier reviewed (approved):
All good from my site (I've fixed a couple of typos directly). Read to merge.

GernotMaier commented:
@VictorBarbosaMartins - please have a look at the conflicts. `data_model/data_model.py` has been renamed to `data_model/meta_data_model.py`.

VictorBarbosaMartins commented:
> @VictorBarbosaMartins - please have a look at the conflicts. `data_model/data_model.py` has been renamed to `data_model/meta_data_model.py`.

Hi Gernot. thanks, I think I solved the conflicts, please take a look once more before merging. Thank you!
